src/WebRTCSession.cpp

```diff
@@ -382,51 +606,134 @@ addDecodeBin(GstElement *webrtc G_GNUC_UNUSED, GstPad *newpad, GstElement *pipe)
+            it != ret.end())
+                std::swap(ret.front(), *it);
+
+        return ret;
+}
+
+}
+
+bool
+WebRTCSession::havePlugins(bool isVideo, std::string *errorMessage)
+{
+        if (!initialised_ && !init(errorMessage))
+                return false;
+        if (!isVideo && haveVoicePlugins_)
+                return true;
+        if (isVideo && haveVideoPlugins_)
+                return true;
+
+        const gchar *voicePlugins[] = {"audioconvert",
+                                       "audioresample",
+                                       "autodetect",
+                                       "dtls",
+                                       "nice",
+                                       "opus",
+                                       "playback",
+                                       "rtpmanager",
+                                       "srtp",
+                                       "volume",
+                                       "webrtc",
+                                       nullptr};
+
+        const gchar *videoPlugins[] = {"opengl", "qmlgl", "rtp", "videoconvert", "vpx", nullptr};
+
+        std::string strError("Missing GStreamer plugins: ");
+        const gchar **needed  = isVideo ? videoPlugins : voicePlugins;
+        bool &havePlugins     = isVideo ? haveVideoPlugins_ : haveVoicePlugins_;
+        havePlugins           = true;
+        GstRegistry *registry = gst_registry_get();
+        for (guint i = 0; i < g_strv_length((gchar **)needed); i++) {
+                GstPlugin *plugin = gst_registry_find_plugin(registry, needed[i]);
+                if (!plugin) {
+                        havePlugins = false;
+                        strError += std::string(needed[i]) + " ";
+                        continue;
                 }
+                gst_object_unref(plugin);
         }
-        return -1;
-}
+        if (!havePlugins) {
+                nhlog::ui()->error(strError);
+                if (errorMessage)
+                        *errorMessage = strError;
+                return false;
+        }
+
+        if (isVideo) {
+                // load qmlglsink to register GStreamer's GstGLVideoItem QML type
+                GstElement *qmlglsink = gst_element_factory_make("qmlglsink", nullptr);
+                gst_object_unref(qmlglsink);
+        }
+        return true;
 }
 
 bool
-WebRTCSession::createOffer()
+WebRTCSession::createOffer(bool isVideo)
 {
-        isoffering_ = true;
+        isOffering_            = true;
+        isVideo_               = isVideo;
+        isRemoteVideoRecvOnly_ = false;
+        videoItem_             = nullptr;
+        haveAudioStream_       = false;
+        haveVideoStream_       = false;
         localsdp_.clear();
         localcandidates_.clear();
-        return startPipeline(111); // a dynamic opus payload type
+        return startPipeline(111, isVideo ? 96 : -1); // dynamic payload types
```

Comment:
What are those numbers?

Reply:
opus and vp8 require you to pick a payload type in the dynamic range (96-127). ie ones not predefined in terms of their encoding type, sampling rate, number of channels etc. I'll add a reference to the comments.

Reply:
Yep, a comment would be perfect, thanks :3